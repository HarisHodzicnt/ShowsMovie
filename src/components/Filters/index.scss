@import '../../styles/variables.scss';

.tabs-filters-page {
  border-right: 2px solid grey;
  .tabs-filters {
    display: flex;
    flex-wrap: wrap;
    justify-content: center;
    .flex-row {
      width: 80%;
      margin: 0 auto;
    }
    button {
      background-color: $app-header-color;
      color: white;
      padding: 5px 20px;
      &:hover {
        color: black !important;
        background-color: transparent !important;
      }
    }
    .tabs {
      margin-top: 50px;
      justify-content: space-evenly;
      width: 80%;
      display: flex;
      .secondary-tab {
        opacity: 0.7;
      }
      height: 40px;
    }
    .filters {
      justify-content: center;
      display: flex;
      flex-wrap: wrap;
      .title {
        margin-top: 50px;
        margin-bottom: 20px;
        font-size: 35px;
        color: #333333;
        text-align: left;
      }
      .flex-row {
        margin-top: 10px;
      }
      .selection {
        margin-top: 0px;
      }
      margin-top: 50px;
    }
    .select {
      background-color: $app-header-color;
      color: white;
      margin-top: 50px !important;
      font-size: 17px;
      svg {
        font-size: 20px;
        margin-left: 5px;
      }
    }
  }
}

@media only screen and (max-width: 950px) {
  .tabs-filters {
    border: none;
    .filters {
      div {
        width: 100%;
      }
    }
  }
}
