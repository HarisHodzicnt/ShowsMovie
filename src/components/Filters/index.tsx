import React, { useCallback, useEffect } from 'react';
import { useHistory } from 'react-router-dom';
import { observer } from 'mobx-react-lite';
import {
  Button,
  FormControl,
  Grid,
  Input,
  InputLabel,
  MenuItem,
  Select,
} from '@material-ui/core';
import SearchIcon from '@material-ui/icons/Search';
import { useStores } from '../../hooks/useStores';
import { MENU_TABS } from '../../constants';
import './index.scss';

const Filters = observer(() => {
  const { movieShowStore } = useStores();
  let history = useHistory();

  useEffect(() => {
    movieShowStore.getAllGenres();
  }, []);

  const handleInputChange = useCallback(
    (
      event: React.ChangeEvent<{
        value: string[] | string;
        id: string;
        name: string;
      }>
    ) => {
      const id = event.target.id ? event.target.id : event.target.name;
      movieShowStore.setFilters({
        ...movieShowStore.filters,
        [id]: event.target.value as string[],
      });
    },
    [movieShowStore.filters]
  );

  const handleSubmit = () => {
    const { genre, name, year } = movieShowStore.filters;
    movieShowStore.getFilteredMoviesShows(genre.join(','), year, name);
  };

  const handleTabClick = (tab: 'Favorites' | 'Movies' | 'Shows') => () => {
    if (tab === 'Favorites') {
      history.push('/favorites');
      return;
    }
    movieShowStore.setActiveTabSelect(tab);
  };

  return (
    <Grid item xs={12} md={4} className="tabs-filters-page">
      <div className="tabs-filters">
        <div className="tabs">
          {MENU_TABS.map((tab: any) => {
            return (
              <Button
                key={tab}
                className={`${
                  tab === movieShowStore.activeTabSelect
                    ? 'active-tab'
                    : 'secondary-tab'
                }`}
                onClick={handleTabClick(tab)}
              >
                {tab}
              </Button>
            );
          })}
        </div>

        <div className="flex-row filters">
          <div className="title">Filters</div>
          <Input
            placeholder={'Filter by name'}
            id="name"
            onChange={handleInputChange}
            value={movieShowStore.filters.name}
            className="flex-row"
            data-testid="filter-name-input"
          />
          <FormControl className="flex-row selection">
            <InputLabel className="genre-label" id="demo-simple-select-label">
              Choose genre
            </InputLabel>

            <Select
              labelId="genre"
              id="genre"
              onChange={handleInputChange as any}
              value={movieShowStore.filters.genre}
              multiple
              input={<Input />}
              name="genre"
            >
              {movieShowStore.allGeneres.map((genre) => (
                <MenuItem value={genre.name} key={genre.id}>
                  {genre.name}
                </MenuItem>
              ))}
            </Select>
          </FormControl>
          <Input
            className="flex-row"
            placeholder={'Filter by year'}
            onChange={handleInputChange}
            id="year"
            value={movieShowStore.filters.year}
          />
          <Button
            onClick={handleSubmit}
            className="select flex-row"
            data-testid="filter-search"
          >
            Search
            <SearchIcon />
          </Button>
        </div>
      </div>
    </Grid>
  );
});

export default Filters;
