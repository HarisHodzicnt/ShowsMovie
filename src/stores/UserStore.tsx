import { makeAutoObservable } from 'mobx';
import { IUser } from '../interfaces/IUsers';
import api from '../services';
export class UserStore {
  public loggedInUser: boolean = false;
  public userData: IUser = { username: '', password: '' };

  constructor() {
    makeAutoObservable(this);
    this.loggedInUser = this.checkLoginToken();
  }
  public loginUser = async ({ userName, password }: any) => {
    await api.login(userName, password);
    this.loggedInUser = true;
    this.userData = { username: userName, password: password };
    window.location.href = '/movies';
  };

  public logoutUser = async () => {
    await api.logout(this.userData.username);
    this.loggedInUser = false;
    window.location.href = '/';
  };

  public checkLoginToken = () => {
    const currentUser: string = localStorage.getItem('currentUser') || '{}';
    let isUserLoggedIn: boolean = false;
    if (currentUser) {
      const allUsers: IUser[] = JSON.parse(
        localStorage.getItem('loginData') || '[]'
      );
      const user = allUsers?.find((user) => user.username == currentUser);
      isUserLoggedIn = !!user;
      this.userData = { username: currentUser, password: '' };
    }
    return isUserLoggedIn;
  };
}
