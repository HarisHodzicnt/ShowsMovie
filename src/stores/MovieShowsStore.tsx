import { makeAutoObservable } from 'mobx';
import { toast } from 'react-toastify';
import { IFilter } from '../interfaces/IFilters';
import { IGenre } from '../interfaces/IGenre';
import { IMovieShow } from '../interfaces/IMovieShow';
import { ISort } from '../interfaces/ISort';
import api from '../services';
import { UserStore } from './UserStore';
export class MovieShowStore {
  public userStore: UserStore = new UserStore();
  public allMovies: IMovieShow[] = [];
  public allShows: IMovieShow[] = [];
  public allFavorites: IMovieShow[] = [];
  public allGeneres: IGenre[] = [];
  public activeTabSelect: 'Movies' | 'Shows' | 'Favorites' = 'Movies';
  public setActiveTabSelect = (state: 'Movies' | 'Shows' | 'Favorites') =>
    (this.activeTabSelect = state);
  public sortByMethod: ISort = { type: 'desc', sortBy: 'release_date' };

  public filters: IFilter = { name: '', genre: [], year: '' };

  constructor() {
    makeAutoObservable(this);
  }

  public setFilters = (filters: IFilter) => (this.filters = filters);

  public getSortedBy = (sortBy: 'name' | 'date') => {
    const type = this.sortByMethod.type === 'asc' ? 'desc' : 'asc';
    const sort: ISort =
      sortBy === 'name'
        ? { type: type, sortBy: 'original_title' }
        : { type: type, sortBy: 'release_date' };
    this.sortByMethod = sort;
    this.getFilteredMoviesShows(
      this.filters.genre.join(','),
      this.filters.year,
      this.filters.name
    );
  };
  public callProperData = () => {
    this.getAllFavorites();
    switch (this.activeTabSelect) {
      case 'Movies':
        this.getAllMovies();
        break;
      case 'Shows':
        this.getAllTvShows();
        break;
    }
  };
  public getAllMovies = async () => {
    const movies = await api.getPopularMovies();
    if (movies) {
      this.allMovies = movies;
      toast.success('getAllMovies', {
        position: toast.POSITION.BOTTOM_CENTER,
      });
    }
  };

  public getAllTvShows = async () => {
    const shows = await api.getPopularTvShows();
    if (shows) {
      this.allShows = shows;
      toast.success('getAllShows', {
        position: toast.POSITION.BOTTOM_CENTER,
      });
    }
  };

  public getAllGenres = async () => {
    const genres = await api.getAllGenres();
    if (genres) {
      this.allGeneres = genres;
      toast.success('getAllGenres', {
        position: toast.POSITION.BOTTOM_CENTER,
      });
    }
  };

  public setFavorite = async (movieOrSerie: IMovieShow) => {
    const favoriteList = JSON.parse(localStorage.getItem('favorites') || '{}');
    const prevFav = await this.getAllFavorites();
    const checkIFExists = prevFav.find(
      (item: IMovieShow) => item.id === movieOrSerie.id
    );
    if (checkIFExists) {
      //remove item from favorites
      const filteredList = prevFav.filter(
        (item: IMovieShow) => item.id !== movieOrSerie.id
      );
      this.allFavorites = [...filteredList];
    } else {
      this.allFavorites = [...prevFav, movieOrSerie];
    }
    localStorage.setItem(
      'favorites',
      JSON.stringify({
        ...favoriteList,
        [this.userStore.userData.username]: this.allFavorites,
      })
    );
    toast.success('Successfully added to the favorites.', {
      position: toast.POSITION.BOTTOM_CENTER,
    });
  };

  public getAllFavorites = async () => {
    const favoriteList = JSON.parse(localStorage.getItem('favorites') || '{}');
    const favoritesCurrentLoggedIn: IMovieShow[] =
      favoriteList[this.userStore.userData.username] || [];
    this.allFavorites = favoritesCurrentLoggedIn;
    return favoritesCurrentLoggedIn;
  };

  public sortFavorites = async (sortBy: ISort) => {
    const favorites = await this.getAllFavorites();
    const sortByTitle = (a: IMovieShow, b: IMovieShow) => {
      const aString = a.original_title || a.name || '';
      const bString = b.original_title || b.name || '';
      return aString.localeCompare(bString);
    };
    const sortByDate = (a: IMovieShow, b: IMovieShow) =>
      new Date(a.release_date).getTime() - new Date(b.release_date).getTime();
    let sortedFavorites: IMovieShow[] = [];
    if (sortBy.sortBy === 'original_title') {
      sortedFavorites = favorites
        .slice()
        .sort((a: IMovieShow, b: IMovieShow) =>
          sortBy.type === 'asc' ? sortByTitle(a, b) : sortByTitle(b, a)
        );
    } else {
      sortedFavorites = favorites
        .slice()
        .sort((a: IMovieShow, b: IMovieShow) =>
          sortBy.type === 'asc' ? sortByDate(a, b) : sortByDate(b, a)
        );
    }
    this.allFavorites = sortedFavorites;
  };

  public getFilteredMoviesShows = async (
    genre: string,
    year: string,
    text: string
  ) => {
    const genreList = genre.split(',');
    //getting genres by name to get ids for API
    const filteredGenre =
      this.allGeneres.filter((item) => genreList.includes(item.name)) || null;
    const genreIds = filteredGenre.map((genre) => genre.id).join(',');

    // choosing what to search movies or shows
    const type =
      this.activeTabSelect === 'Movies'
        ? 'movie'
        : this.activeTabSelect === 'Shows'
        ? 'tv'
        : '';

    //sort by, original_title.asc
    const sortBy = `${this.sortByMethod.sortBy}.${this.sortByMethod.type}`;
    const filteredMovieOrShow = await api.getFilteredMoviesShows(
      genreIds,
      year,
      text,
      type,
      sortBy
    );
    if (filteredMovieOrShow) {
      if (this.activeTabSelect === 'Movies') {
        this.allMovies = filteredMovieOrShow;
      } else if (this.activeTabSelect === 'Shows') {
        this.allShows = filteredMovieOrShow;
      }
      toast.success('getFilteredMoveis', {
        position: toast.POSITION.BOTTOM_CENTER,
      });
    }
  };
}
