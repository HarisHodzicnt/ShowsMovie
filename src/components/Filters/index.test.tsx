import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import axios from 'axios';
import Filters from '.';
import { useStores } from '../../hooks/useStores';
import { MovieShowStore } from '../../stores/MovieShowsStore';

jest.mock('../../hooks/useStores');
jest.mock('axios');
const mockGet = jest.fn();
const mockPost = jest.fn();
axios.get = mockGet;
axios.post = mockPost;

test('Test MovieShowCard component', () => {
  useStores.mockReturnValue({
    movieShowStore: new MovieShowStore(),
  });
  const card = render(<Filters />);
  expect(card.container).toMatchSnapshot();
  userEvent.type(screen.getByTestId('filter-name-input'), 'Dragon');
  userEvent.click(screen.getByTestId('filter-search'));
  expect(mockGet).toBeCalledWith(
    'https://api.themoviedb.org/3/search/movie?api_key=0358321151dc20553c013d0dd7246aa6&with_genres=&year=&query=Dragon&sort_by=release_date.desc'
  );
});
