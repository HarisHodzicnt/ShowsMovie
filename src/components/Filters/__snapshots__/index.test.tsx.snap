// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test MovieShowCard component 1`] = `
<div>
  <div
    class="MuiGrid-root tabs-filters-page MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4"
  >
    <div
      class="tabs-filters"
    >
      <div
        class="tabs"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text active-tab"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Movies
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text secondary-tab"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Shows
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text secondary-tab"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Favorites
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="flex-row filters"
      >
        <div
          class="title"
        >
          Filters
        </div>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline flex-row"
          data-testid="filter-name-input"
        >
          <input
            class="MuiInputBase-input MuiInput-input"
            id="name"
            placeholder="Filter by name"
            type="text"
            value=""
          />
        </div>
        <div
          class="MuiFormControl-root flex-row selection"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root genre-label MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink="false"
            id="demo-simple-select-label"
          >
            Choose genre
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          >
            <div
              aria-haspopup="listbox"
              aria-labelledby="genre genre"
              class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
              id="genre"
              role="button"
              tabindex="0"
            >
              <span>
                â€‹
              </span>
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput"
              name="genre"
              tabindex="-1"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSelect-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline flex-row"
        >
          <input
            class="MuiInputBase-input MuiInput-input"
            id="year"
            placeholder="Filter by year"
            type="text"
            value=""
          />
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text select flex-row"
          data-testid="filter-search"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Search
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
